File: weirdgrids/java/GridFiller.java
import java.util.Random;

public class GridFiller{
    private static char[][] canvas;
    private static volatile boolean[] flags;
    private static Random randomizer = new Random();

    private static final char[] BINARY = {'0', '1'};

    private static class Artist implements Runnable{
        private int row;
        private int duration;

        public Artist(int row, int duration){
            this.row = row;
            this.duration = duration;
        }

        private synchronized void fillGrid(){
            int randomIndex = randomizer.nextInt(canvas[row].length);
            int randomBit = randomizer.nextInt(2);

            System.out.println("Random index: (" + row + ", " + randomIndex + ")");

            canvas[row][randomIndex] = BINARY[randomBit];
        }
    
        @Override
        public void run(){
            for(int i = 0; i < duration; i++){
                fillGrid();

                try{
                    Thread.sleep(1000);
                } catch(InterruptedException ie){
                    System.err.println("Artist InterruptedException!");
                }
            }
        }
    }

    private static boolean allDone(){
        for(int i = 0; i < flags.length; i++){
            if(!flags[i]){
                return false;
            }
        }

        return true;
    }

    private static class Animator implements Runnable{
        @Override
        public void run(){
            while(!allDone()){
                System.out.println("Animator will try to draw...TODO LOLzzz!!");
                try{
                    Thread.sleep(1000);
                } catch(InterruptedException ie){
                    System.err.println("Animator InterruptedException!");
                }
            }
        }
    }

    private static String gridObject(){
        StringBuilder builder = new StringBuilder();

        for(int i = 0; i < canvas.length; i++){
            builder.append(canvas[i]);
        }

        return builder.toString();
    }

    public static void main(String[] args) throws Exception{
        if(args.length != 3){
            System.out.println("Usage: java GridFiller <grid width> <grid height> <duration>");
            System.exit(1);
        }

        int gridWidth = Integer.parseInt(args[0]);
        int gridHeight = Integer.parseInt(args[1]);
        int duration = Integer.parseInt(args[2]);

        canvas = new char[gridHeight][gridWidth];

        for(int row = 0; row < gridHeight; row++){
            for(int col = 0; col < gridWidth; col++){
                canvas[row][col] = '0';
            }
        }

        flags = new boolean[gridHeight];

        for(int i = 0; i < gridHeight; i++){
            Thread t = new Thread(new Artist(i, duration));
            t.start();
        }

        Thread t = new Thread(new Animator());
        t.start();
    }
}

File: weirdgrids/gridfiller.py
import gevent
import os
import random
import sys
import time

from gevent import Greenlet
from gevent.lock import Semaphore

chargrid = None
semaphore_lock = None
flags = None

def grid_join():
    return "\n\r".join(["".join(ss) for ss in chargrid])

def fill_grid(row, duration):
    for i in range(duration):
        random_index = random.choice(range(len(chargrid[row])))
        random_bit = random.choice(("1", "0"))
    
        semaphore_lock.acquire()
        chargrid[row][random_index] = random_bit
        semaphore_lock.release()
        #time.sleep(1)
        gevent.sleep(1)

    print str(row) + " is done."
    flags[row] = True

def all_true():
    for b in flags:
        if not b:
            return False
    
    return True

def animator():
    while not all_true():
        os.system("clear")
        sys.stdout.write(grid_join() + "\r")
        sys.stdout.flush()
        #print "Animator will draw..."
        #time.sleep(1)
        gevent.sleep(1)

if __name__ == "__main__":
    if len(sys.argv) != 4:
        print "Usage: python " + sys.argv[0] + " <grid width> <grid height> <duration>"
        exit(1)
    
    width_cols = int(sys.argv[1])
    height_rows = int(sys.argv[2])
    duration = int(sys.argv[3])

    chargrid = [[' ' for i in range(width_cols)] for j in range(height_rows)]
    flags = [False for j in range(height_rows)]
    semaphore_lock = Semaphore()

    draw_threads = []

    for i in range(height_rows):
        artist = gevent.spawn(fill_grid, i, duration)
        draw_threads.append(artist)
    
    draw_threads.append(gevent.spawn(animator))
    gevent.joinall(draw_threads)
